### PR TITLE
google-cloud-cpp: 0.14.0 -> 1.32.1

### PR DESCRIPTION
**Pull request by cpcloud**
###### Motivation for this change

Update google-cloud-cpp

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by cpcloud** [comment]
This PR depends on #143967.

File: pkgs/development/libraries/google-cloud-cpp/default.nix [hunk 1]
```diff
@@ -1,76 +1,88 @@
...
   pname = "google-cloud-cpp";
-  version = "0.14.0";
+  version = "1.32.1";
 
   src = fetchFromGitHub {
     owner = "googleapis";
     repo = "google-cloud-cpp";
     rev = "v${version}";
-    sha256 = "15wci4m8h6py7fqfziq8mp5m6pxp2h1cbh5rp2k90mk5js4jb9pa";
+    hash = "sha256:0g720sni70nlncv4spm4rlfykdkpjnv81axfz2jd1arpdajm0mg9";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/google-cloud-cpp/default.nix:37]:
```suggestion
    sha256 = "0g720sni70nlncv4spm4rlfykdkpjnv81axfz2jd1arpdajm0mg9";
```

File: pkgs/development/libraries/google-cloud-cpp/default.nix [hunk 2]
```diff
@@ -1,76 +1,88 @@
...
 
-  buildInputs = [ curl crc32c googleapis-cpp-cmakefiles grpc protobuf libnsl ];
-  nativeBuildInputs = [ clang-tools cmake pkg-config doxygen ];
-
-  outputs = [ "out" "dev" ];
+  postInstall = ''
+    # for some reason there are already includes in $dev at this point which
+    # causes a problem for the fixup hooks that move stuff from out into dev
+    # so we remove the includes in dev so they can be filled in by the hooks
+    rm -rf $dev/include
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/google-cloud-cpp/default.nix:44]:
```suggestion
    rm -r $dev/include
```

File: pkgs/development/libraries/google-cloud-cpp/default.nix [hunk 3]
```diff
@@ -1,76 +1,88 @@
...
+  '';
 
   postPatch = ''
-    sed -e 's,https://github.com/nlohmann/json/releases/download/v3.4.0/json.hpp,file://${_nlohmann_json},' \
-    -i cmake/DownloadNlohmannJson.cmake
+    sed \
+      -e 's,https://github.com/googleapis/googleapis/archive/${googleapisRev}.tar.gz,file://${googleapis},' \
+      -i external/googleapis/CMakeLists.txt
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/google-cloud-cpp/default.nix:50]: Why not use substituteInPlace?

File: pkgs/development/libraries/google-cloud-cpp/default.nix [hunk 4]
```diff
@@ -1,76 +1,88 @@
...
+    curl
+    grpc
+    gbenchmark
+    gmock
+    gtest
+    nlohmann_json
+    protobuf
+  ];
+
+  nativeBuildInputs = [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/google-cloud-cpp/default.nix:66]: native usually goes before buildInputs.

File: pkgs/development/libraries/google-cloud-cpp/default.nix (hunk 2 above)
Comment (cpcloud) [reply pkgs/development/libraries/google-cloud-cpp/default.nix:44]: This isn't actually necessary any more. I'll remove the `postInstall` entirely.

File: pkgs/development/libraries/google-cloud-cpp/default.nix (hunk 3 above)
Comment (cpcloud) [reply pkgs/development/libraries/google-cloud-cpp/default.nix:50]: That's what was here before, I can change it.

File: pkgs/development/libraries/google-cloud-cpp/default.nix (hunk 4 above)
Comment (cpcloud) [reply pkgs/development/libraries/google-cloud-cpp/default.nix:66]: Fair enough. I'll move it.

**Comment by cpcloud** [comment]
@SuperSandro2000 Any chance you can give this another review when you get a moment?

File: pkgs/development/libraries/google-cloud-cpp/default.nix [hunk 5]
```diff
@@ -1,76 +1,81 @@
...
+    crc32c
+    curl
+    grpc
+    gbenchmark
+    gtest
+    nlohmann_json
+    protobuf
+  ];
+
+  ninjaFlags = [ "-v" ];
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/google-cloud-cpp/default.nix:65]: Is this left from testing?

File: pkgs/development/libraries/google-cloud-cpp/default.nix [hunk 6]
```diff
@@ -1,76 +1,81 @@
...
-  outputs = [ "out" "dev" ];
-
   postPatch = ''
-    sed -e 's,https://github.com/nlohmann/json/releases/download/v3.4.0/json.hpp,file://${_nlohmann_json},' \
-    -i cmake/DownloadNlohmannJson.cmake
+    substituteInPlace \
+      external/googleapis/CMakeLists.txt \
+      --replace \
+      "https://github.com/googleapis/googleapis/archive/${googleapisRev}.tar.gz" \
+      "file://${googleapis}"
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/google-cloud-cpp/default.nix:44]:
```suggestion
    substituteInPlace external/googleapis/CMakeLists.txt \
      --replace "https://github.com/googleapis/googleapis/archive/${googleapisRev}.tar.gz" "file://${googleapis}"
```

usually we format it like this.

**Comment by SuperSandro2000** [comment]
other than the two minor things above this LGTM

File: pkgs/development/libraries/google-cloud-cpp/default.nix (hunk 5 above)
Comment (cpcloud) [reply pkgs/development/libraries/google-cloud-cpp/default.nix:65]: No, but I can remove it.

**Comment by cpcloud** [comment]
@SuperSandro2000 Fixed, this should be good to go now. I have a follow up to this that will enable a large portion of the test suite for this package.

**Comment by cpcloud** [comment]
@SuperSandro2000 Can this be merged?

**Comment by mkg20001** [comment]
That's quite a lot of commits, that'll clobber history too much. You can actually just squash everything into one upgrade commit.

**Comment by cpcloud** [comment]
@mkg20001 Right, sorry, meant to do that.

**Comment by cpcloud** [comment]
@SuperSandro2000 @mkg20001 Rebased and squashed.

**Comment by cpcloud** [comment]
Thanks @mkg20001 !

File: pkgs/development/libraries/google-cloud-cpp/default.nix [hunk 7]
```diff
@@ -1,76 +1,76 @@
...
-    postPatch = ''
-      sed -e 's,https://github.com/googleapis/googleapis/archive/9c9f778aedde02f9826d2ae5d0f9c96409ba0f25.tar.gz,file://${googleapis},' \
-      -i CMakeLists.txt
-    '';
-  };
-  _nlohmann_json = fetchurl {
-    url = "https://github.com/nlohmann/json/releases/download/v3.4.0/json.hpp";
-    sha256 = "0pw3jpi572irbp2dqclmyhgic6k9rxav5mpp9ygbp9xj48gnvnk3";
+    rev = googleapisRev;
+    hash = "sha256:1xrnh77vb8hxmf1ywqsifzd39kylhbdyah0b0b9bm7nw0mnahssl";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/google-cloud-cpp/default.nix:25]: Nit: This should use sha256 or an Sri hash, not a mixture of both.

**Pull request merged**